README.md
# Teletext EPG grabber

This EPG grabber or "scraper" allows extracting TV programme listings in
[XMLTV format](http://wiki.xmltv.org) from teletext pages as broadcast by
most European TV networks. The grabber works by reading teletext pages via
a TV card from all individual networks, then extracting programme title
and starting times from programme tables (e.g. in Germany usually pages
301-304) and finally adding description texts from pages which are
referenced in the overviews.

The grabber can be used stand-alone, but is best used as part of the
Nextview EPG grabber [nxtvepg](http://nxtvepg.sourceforge.net) (version
2.9.0 or later), which can also be used for immediately browsing the
programme listings in the generated XMLTV files. The grabber is integrated
into the nxtvepg acquisition control and can be activated via the
configuration menus. (Version 2.9 of nxtvepg still defaults to acquisition
via Nextview EPG, but that service is defunct now; use nxtvepg version 3.0
or later for optimized support.)

Since the grabber has to read each individual network's teletext stream and
many networks use sub-pages at least for descriptions, the process if done
sequentially takes relatively long, namely at least 90-120 seconds per network
for grabbing all data including description texts.  However, when capturing
from DVB, it is possible to capture from all channels sharing a transponder,
which in average speeds up acquisition by more than factor 5 (see description
of script `capall.sh` below). There is also a fast scan mode which requires
only 10-20 seconds per network in average, however in this mode overview and
descriptions may be incomplete if sub-pages are used by the respective network.

For now, only nation-wide German networks are supported well. For other
networks, the parser would need to be adapted to different page formats. That's
because almost every network formats tables, dates, times and descriptions
slightly differently, so that the parser will have trouble locating the EPG
data among all the advertisements and other content in teletext.

# Latest news: March/2021: Release of version 2.3

Development was resumed with release 2.0 after a 9 year gap. The first release
mainly focused on enhancing working with digital TV capture cards (e.g. by
adapting the `capall.pl` example automation script). Also tools have been
updated (e.g. removed the need for the Boost regex library by requiring C++11)
and the parser was slightly enhanced and adapted to changes in teletext content
by networks.

Subsequent releases provided further enhancements and fixes in teletext page
parser and tools (e.g. DVB channel tuning), as well as interface adaption for
integration in nxtvepg. See file "changelog" for details.

# Installation

If you got the Debian package (or created one yourself using shell script
`mkdebian.pl`), you can install the executable and manual page by typing:

```console
    dpkg -i PACKAGE
```

Alternatively, you can compile and install the package from sources directly by
typing `make install prefix=/usr/local` (optionally replacing `/usr/local` with
another directory which contains bin/ and man/, where to install the executable
and manual page respectively.)

For use with nxtvepg: The grabber is actually compiled into the nxtvepg
executable: Both nxtvepg release 2.9.0 and 3.0.0 include the current latest
grabber release 2.3. See
[nxtvepg grabber documentation](http://nxtvepg.sourceforge.net/ttx_grab.html).

# Dependencies

This grabber captures EPG data from live TV broadcast, therefore a TV capture
device (e.g. a DVB capture device) is required.

When not used from within [nxtvepg](http://nxtvepg.sourceforge.net), the grabber
requires the [ZVBI library](http://zapping.sourceforge.net/ZVBI/index.html) for
capturing teletext from a TV capture device. The library is available for Linux
and BSD only. Most Linux distributions provide the library as a package (e.g.
for Debian, Ubuntu et.al. use `apt-get install libzvbi-dev`).

For building the executable, a C++ compiler supporting release C++11 or later
is required.

Note up to version 1.x of this package [Perl](https://www.perl.org/) and the
[Video::ZVBI](https://metacpan.org/pod/Video::ZVBI) Perl extension module were
used for implementation (actually that extension was developed for this
purpose). Due to increased size and complexity of the implementation however,
C++ is used instead since version 2.0, so these dependencies are moot.

# Usage

The following is a usage example: First the channel is switched to network ARD
using `mplayer` (you could use other tools such as `dvb5-zap` for this purpose,
depending on your DVB setup.) Then the grabber is run: With the shown options,
it will capture pages 300-399 from DVB stream 104 (which is the teletext PID of
network "Das Erste", formerly known as "ARD"; you can obtain the PID using tool
`tune_dvb` in this package). Capturing will stop after all sub-pages of the page
range have been acquired. Afterwards the grabber extracts EPG data and generates
an XMLTV file which is written to *stdout*, which is redirected to a file in the
example.  Finally the XMLTV file is loaded into nxtvepg for browsing the
programme schedules. (Any other XMLTV-capable browser will work, too.)

```console
  mplayer dvb://"Das Erste" &
  ./tv_grab_ttx -dvbpid 104 -page 301-399 > ttx-ARD.xml
  nxtvepg ttx-ARD.xml
```

For additional options invoke the grabber with command line option "`-help`".

Using the above will produce a separate XMLTV file per channel, which is not
practical for browsing. Use script `merge.pl` for merging XML files of all
channels into a single XMLTV file. When using the grabber from within nxtvepg,
the merge can be configured to be done automatically.

See script `capall.pl` for an example how to automate capturing and merging EPG
data from multiple channels. (Note this script requires `make tune_dvb` for
tuning into channels and you will have to edit it for your `channels.conf` path
and channel names used therein.) In particular the script speeds-up grabbing
significantly by running `tv_grab_ttx` multiple times concurrently for all DVB
channels sharing a transponder.

# Copyright

**Copyright 2006-2011, 2020-2021 by T. Zoerner**

This program is free software; you can redistribute it and/or modify
it under the terms of the
[GNU General Public License](http://www.fsf.org/copyleft/gpl.html)
as published by the [Free Software Foundation](http://www.fsf.org/).
This program is distributed in the hope that it will be useful,
but **without any warranty**; without even the implied warranty of
merchantability or fitness for a particular purpose.

**Content copyright**: Please note that content providers do hold a copyright
on the programme schedules which can be extracted by means of this software.
The data is free for personal use, but must not be publicly redistributed (e.g.
made available on the Internet) without prior permission.


changelog
libxmltv-ttx (2.3) unstable; urgency=medium

  * Filter overlapping TV programmes from XMLTV output.
  * Filter TV programmes too far in the future (likely result of misdetection)
  * Fixed a bug that lead to gaps in TV schedule output for TV programmes
    after midnight.
  * Added the teletext page number of the teletext overview page it originates
    from as comment for each <programme> element in XMLTV output.
  * Added a "date" attribute to the <tv> tag, specifying time of acquisition.
  * Added translation for Italian month & weekday names.
  * Extended the time_dvb tool to support "mplayer" channels.conf format
    and tuning of DVB-T and DVB-S channels.
  * Added auto-detection for the teletext PID to the tune_dvb tool.
  * Source code was adapted for compiling with gcc version 9.

 -- T. Zoerner <tomzo@users.sourceforge.net>  Sun, 21 Mar 2021 10:23:29 +0200

libxmltv-ttx (2.2) unstable; urgency=medium

  * Adapted internal C interface used by nxtvepg branch "devel_3_0_0" for
    allowing capturing concurrently from multiple DVB channels
  * Bugfix "capall" to use TTX page numbers given in config
  * Enhanced robustness of merge script (e.g. duplicate channel names)
  * Enhanced parsing of series indices in programme titles

 -- T. Zoerner <tomzo@users.sourceforge.net>  Tue, 22 Dec 2020 14:41:00 +0200

libxmltv-ttx (2.1) unstable; urgency=medium

  * Replaced use of boost_regex library with C++11 std::regex
  * Changed default from analog to digital TV capture device
  * Added command line option "-pgstat" for monitoring capture progress
  * Added optional tool for tuning DVB channels in util/ directory
  * Enhanced example capture script capall.pl: receive concurrently
    from channels sharing a transponder; can speed-up overall duration
    of acquisition significantly (e.g. by factor 4)
  * Minor parser enhancements

 -- T. Zoerner <tomzo@users.sourceforge.net>  Tue, 30 Jun 2020 20:24:00 +0200

libxmltv-ttx (2.0) unstable; urgency=medium

  * Complete re-implemenation in C++
  * Added support for new "HD" feature attribute
  * Updated XMLTV output to DTD 0.5.56: "bilingual", "HDTV"
  * Map "audio-descriptive" feature flag to bilingual
  * Remove redundancy from descriptions on multiple sub-pages
  * Improved fuzzy-matching on dates on description page dates
  * Include description text even upon missing or unreadable date
    of the page is referenced by a single programme
  * Improve separation between programme titles and sub-titles in
    overwiew pages using lin-breaks found on description pages
  * Bug-fix end-time parser for ARTE
  * Fixed TTX page ref matching: avoid false matches
  * Improved rubustness of cast table parser and support more table formats
  * Implementation switched from Perl to C++ (code design improvement)
  * Improved merge to avoid losing descriptions or feature flags

 -- T. Zoerner <tomzo@users.sourceforge.net>  Sun, 09 Jan 2011 19:38:01 +0100

libxmltv-ttx (1.1) unstable; urgency=low

  * Added documentation inside of tv_grab_ttx.pl (man page in POD format)
  * Included functionality of previous "cap.pl" into tv_grab_ttx.pl
  * Renamed grabber script from ttx_grab.pl into tv_grab_ttx.pl
  * Added script to generate Debian package for installation

  * Grabber parses and reformats "cast" tables on description tables

 -- T. Zoerner <tomzo@users.sourceforge.net>  Sun, 05 Oct 2008 21:35:00 +0200

libxmltv-ttx (1.0) unstable; urgency=low

  * Adapted date parsing for changes in BR3 and others
  * Improved & extended attribute parsing on overview pages
  * Improved channel display name parsing;
    fall back to using channel name on failure instead of "???"

 -- T. Zoerner <tomzo@users.sourceforge.net>  Sat, 08 Mar 2008 00:00:00 +0200


capall.pl
#!/usr/bin/perl -w
#
#  This is a demo how to capture EPG data from multiple channels: The
#  script will tune all the channels given in the list. For each channel
#  it will capture teletext pages in the given ranges for 90 seconds, then
#  try scraping TV schedules from them and write the result to a file in
#  XMLTV format (while merging the new EPG data with the one present in
#  the file). The latter is done in parallel for all channels sharing a
#  transponder. At the end all XML files in the current directory are
#  merged into one using helper script merge.pl
#
#  This script does not parse command line parameters. Please edit the
#  configuration variables at the top of the script as needed.
#
#  Tool "util/tune_dvb.c" is used for tuning the DVB front-end: use "make
#  tune_dvb" for building the executable. Note the tool only supports
#  "channels.conf" tables mplayer/totem/xine/kaffeine and VDR formats.
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
#  Copyright 2006-2008,2020-2021 by T. Zoerner (tomzo at users.sf.net)
#

use strict;

#
# Configuration variables
#
my $channels_conf = "$ENV{HOME}/.mplayer/channels.conf";
my $device = "/dev/dvb/adapter0/demux0";
my $def_duration = 90;
my $def_pages = "300-399";

my @Nets = (
   ["Das Erste", "ard"],
   ["ZDF", "zdf"],
   ["arte", "arte", "300-699"],    # today on 300-300, next 600-699
   ["KiKA", "kika"],
   ["3sat", "3sat"],
   ["ONE", "ard-one"],
   ["WDR", "wdr"],
   ["TELE 5", "tele5"],
   ["SAT.1", "sat1"],
   ["RTL", "rtl"],
   ["RTLplus", "rtl-plus"],
   ["ProSieben", "pro7"],
   ["RTLZWEI", "rtl2"],
   ["VOX", "vox"],
   ["VOXup", "vox-up"],
   ["sixx", "sixx"],
   #["ServusTV", "servus"], # no teletext
   ["DMAX", "dmax"],
   #["MTV", "mtv"],    # no teletext
   ["MDR", "mdr"],
   ["NDR FS MV", "ndr"],
   ["SWR Fernsehen RP", "swr"],
   ["BR Fernsehen", "br3"],
   ["hr-fernsehen", "hr3", undef, 120],
   ["rbb Berlin", "rbb"],
   ["ARD-alpha", "bralpha"],
   ["phoenix", "phoenix"],
   ["kabel eins", "kabel1"],
   ["SUPER RTL", "srtl"],
   ["Eurosport 1", "eurosport"],
   #["TV 5 Monde", "tv5"],       # no teletext
   ["ntv", "ntv", "500-599"],
   #["CNN", "cnni", "200-299"],  # no teletext
);

# catch signal "INT" for removing "*.tmp" files created for tv_grab_ttx output redirection
my @SigTmps = ();
sub sig_handler {
   my ($sig) = @_;
   print "Caught a SIG$sig - shutting down\n";
   unlink($_) foreach (@SigTmps);
   @SigTmps = ();
   exit(1);
};
$SIG{'INT'}  = \&sig_handler;

while (@Nets) {
   # get name of first channel that's not done yet
   my $chan = $Nets[0][0];

   # query for list of all channels sharing the transponder of this channel
   my @tids = ();
   if (open(TUNE, "./tune_dvb -c \"$channels_conf\" -P \"$chan\"|")) {
      while (<TUNE>)
      {
         if (/(\d+)\t(.*)/) {
            # search returned channel name in the "Nets" list
            for (my $idx = 0; $idx <= $#Nets; $idx++) {
               if ($Nets[$idx][0] eq $2) {
                  # match -> schedule this channel in this iteration
                  push(@tids, [$1, $Nets[$idx][1], $Nets[$idx][2], $Nets[$idx][3]]);
                  # remove channel from "Nets" list as it is done after this iteration
                  splice(@Nets, $idx, 1);
                  last;
               }
            }
         }
      }
   }
   if (@tids) {
      # tuning option for analog TV card (obsolete)
      #system "v4lctl -c $device setstation \"$chan\"";

      # tune the transponder of the selected channels
      # FIXME open does not return error when tuning fails
      if (open(TUNE, "|./tune_dvb -c \"$channels_conf\" \"$chan\"")) {

         print "Capturing from $chan (".scalar(@tids)." PIDs)\n";
         my @pids = ();
         # run tv_grab_ttx for each of the channels as background process
         foreach (@tids) {
            my ($tid, $fname, $pages, $duration) = @$_;
            $pages = $def_pages if !defined $pages;
            $duration = $def_duration if !defined $duration;
            print "- capturing $fname (PID $tid, pages $pages, duration $duration)\n";
            my $out_file = "ttx-$fname.xml";
            push @SigTmps, "${out_file}.tmp";

            my $pid = fork();
            die "fork: $!\n" unless defined $pid;

            if ($pid == 0) {
               # child process: build the grabber command line & exec tv_grab_ttx
               # - capture into temporary file
               # - merge output with previous XML file of same name, if one exists
               open(STDOUT, '>', "${out_file}.tmp") or die "Can't redirect STDOUT >${out_file}.tmp: $!";

               my @cmd  = ("./tv_grab_ttx", "-dev", $device,
                                            "-duration", $duration,
                                            "-page", $pages);
               push(@cmd, "-dvbpid", $tid);
               push(@cmd, "-merge", $out_file) if -e $out_file && !-z $out_file;
               exec @cmd;
               die "exec @cmd: $!\n";
            }
            else {
               # parent process: remember the child PID for waiting below
               push @pids, [$pid, $out_file];
            }
         }
         # wait for all tv_grab_ttx processes to be done
         foreach my $pid (@pids) {
            if (waitpid($pid->[0], 0) == $pid->[0]) {
               my $stat = $?;
               my $out_file = $pid->[1];

               # check exit code of tv_grab_ttx: discard output if not zero
               if (($stat >> 8) == 0) {
                  if (-e "${out_file}.tmp") {
                     # replace XML file for that channel with "*.tmp" file created by tv_grab_ttx output
                     rename("${out_file}.tmp", $out_file) || warn "error renaming/replacing $out_file: $!\n";
                  }
                  else {
                     warn "missing output file ${out_file}.tmp\n";
                  }
               }
               else {
                  warn "tv_grab_ttx failed for $out_file\n";
                  unlink "${out_file}.tmp";
               }
            }
         }
         @SigTmps = ();
         close(TUNE);
      }
      else {
         #print "Failed to tune $chan\n";  # printed by sub-command
      }
   }
   else {
      print "No teletext PID for $chan\n";
      splice(@Nets, 0, 1);
   }
}

print "Finally merging all captured XMLTV files into tv.xml\n";
system "./merge.pl ttx-*.xml > tv.xml";
